lib/dry/core/memoizable.rb

```diff
@@ -12,7 +12,7 @@ def memoize(*names)
 
         def new(*)
           obj = super
-          obj.instance_variable_set(:'@__memoized__', MEMOIZED_HASH.dup)
+          obj.instance_eval { @__memoized__ = MEMOIZED_HASH.dup }
```

Comment:
Could this be separated? This will have a performance impact for sure so it'd be better to have a dedicated `new` method for Object and BasicObject descendants.